pkg/deploy/deploy.go

```diff
@@ -208,51 +210,111 @@ func deployComponentsParallel(ctx context.Context, components []graph.SortedComp
 	return errs
 }
 
-func deployComponent(ctx context.Context, component graph.SortedComponent, clientSet ClientSet, apis api.APIs, opts DeployConfigsOptions) []error {
+type componentDeployer struct {
+	lock             sync.Mutex
+	graph            graph.ConfigGraph
+	clients          ClientSet
+	resolvedEntities entityMap
+	apis             api.APIs
+}
+
+func (c *componentDeployer) deploy(ctx context.Context) []error {
 	var errs []error
 
-	entityMap := newEntityMap(apis) //entityMap is only used to when resolving parameter references, and configs that reference each other are in the same component; no global entity map is needed
+	errChan := make(chan error)
+	for c.graph.Nodes().Len() != 0 {
+		roots := graph.Roots(c.graph)
 
-	g := component.Graph
-	sortedNodes := component.SortedNodes
+		for _, root := range roots {
+			go func(node graph2.Node) {
+				errChan <- c.deployNode(ctx, node)
```

Comment:
I think you can run into the same race condition as the go routines one level up in DeployParallel did? 
You pass the same ctx pointer into the different `deployNode` methods, which modify it (line 255)

If you create a copy of the context already with the Coordinate value in it and pass that into the go func (compare line 194) you'd be safe from that. And you have the coordinate available here from the root ConfigNode

Reply:
I can't see how it would, as the WithValue is implemented as follows:
```go
func WithValue(parent Context, key, val any) Context {
	if parent == nil {
		panic("cannot create context from nil parent")
	}
	if key == nil {
		panic("nil key")
	}
	if !reflectlite.TypeOf(key).Comparable() {
		panic("key is not comparable")
	}
	return &valueCtx{parent, key, val}
}
```
> You pass the same ctx pointer into the different deployNode methods, which modify it (line 255)
It should not matter if we call it before or after as the parent is never modified.
To be sure, I tested it: https://go.dev/play/p/woxG3YbTiNZ


However, I think it makes sense to move it outside, as it is IMO a bit cleaner, and we can use the ConfigNode directly, meaning less unnecessary casts
It helps us generally and is a bit clearer imo.

Reply:
You're right, passing the ctx on, you're modifying a different context. 
The forgot the original issue had a different setup (and I forgot we had another issue in the logger further down)